src/mono/mono/metadata/marshal.c

```diff
@@ -3069,19 +3089,65 @@ mono_marshal_set_callconv_from_modopt (MonoMethod *method, MonoMethodSignature *
 		gboolean required;
 		MonoType *cmod_type = mono_type_get_custom_modifier (sig->ret, i, &required, error);
 		mono_error_assert_ok (error);
-		MonoClass *cmod_class = mono_class_from_mono_type_internal (cmod_type);
-		if ((m_class_get_image (cmod_class) == mono_defaults.corlib) && !strcmp (m_class_get_name_space (cmod_class), "System.Runtime.CompilerServices")) {
-			const char *cmod_class_name = m_class_get_name (cmod_class);
-			if (!strcmp (cmod_class_name, "CallConvCdecl"))
-				csig->call_convention = MONO_CALL_C;
-			else if (!strcmp (cmod_class_name, "CallConvStdcall"))
-				csig->call_convention = MONO_CALL_STDCALL;
-			else if (!strcmp (cmod_class_name, "CallConvFastcall"))
-				csig->call_convention = MONO_CALL_FASTCALL;
-			else if (!strcmp (cmod_class_name, "CallConvThiscall"))
-				csig->call_convention = MONO_CALL_THISCALL;
+		mono_marshal_set_callconv_for_type (cmod_type, csig, NULL);
+	}
+}
+
+static void
+mono_marshal_set_callconv_from_unmanaged_callconv_attribute (MonoMethod *method, MonoMethodSignature *csig, gboolean *skip_gc_trans /*out*/)
+{
```

Comment:
A couple of general comments about this method:
1. As written now, it's not usable by the AOT compiler.
2. Can you use `mono_reflection_create_custom_attr_data_args_noalloc` - that should avoid creating managed objects and will instead just give you C structures.  (It's possible it doesn't handle `Type[]` yet, in which case we should fix that - as a separate PR.)
3. There's a pattern to cleanup `MonoCustomAttrInfo` values:
   ```
      	if (!cinfo->cached)
		mono_custom_attrs_free (cinfo);
   ```
4. If the managed methods can't be avoided, the portion of the function that uses handles should be:
  a. pulled out into its own function, 
  b. that function should be single-exit and the body of it should be wrapped in `HANDLE_FUNCTION_ENTER()`/`HANDLE_FUNCTION_RETURN_{,VAL,OBJ}()`

Reply:
> It's possible it doesn't handle `Type[]`

This is what I hit when I had tried to use the noalloc version. `mono_reflection_create_custom_attr_data_args_noalloc` doesn't handle arrays - `load_cattr_value` requires being given an out obj for arrays.

Reply:
I see.  well we should take the managed object version (with the other two cleanups I mentioned) then, and open a GH Issue to improve `_noalloc` and get the AOT compiler support in.

Reply:
Filed https://github.com/dotnet/runtime/issues/52977